Method: 
1) original code
2) deleted common english words and .,'" 
3) 2+ deleted english women, men names 
4) 2+ tried making a probability of not only True/False but with the number of word occurance
4) 4+ added 1 to the zeros for laplace 


In [571]:
import nltk
nltk.download('movie_reviews')
nltk.download('stopwords')
nltk.download('names')
from nltk.corpus import movie_reviews
import random 

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package names to /root/nltk_data...
[nltk_data]   Package names is already up-to-date!


In [572]:
#category: each review as positive or negative 
documents = [(list(movie_reviews.words(fileid)),category) for category in movie_reviews.categories() for fileid in movie_reviews.fileids(category)]
documents[0][1]
# documents : [wordslist, category] 


'neg'

In [573]:
a = documents[0] #first document 
a[0] #this is the list of words 
a[1] #the second element is the category of the movie review 
a[0][0] #the first word of the list of words 

'plot'

In [0]:

random.shuffle(documents) #shuffle the document elements 

In [575]:
all_words = nltk.FreqDist(w.lower() for w in movie_reviews.words())
word_features = list(all_words)[:2000] #list all the words in movie_reviews and use 2000 words as word features 
word_features[:10]
#len(word_features)

['plot', ':', 'two', 'teen', 'couples', 'go', 'to', 'a', 'church', 'party']

In [576]:
all_words['plot'] #a dic for each word in every movie review 
#shows how many times the word has been used  


1513

In [0]:
def document_features(document):
  document_words =set(document) #unique words of the document 
  features = {}
  for word in word_features:
    features[word] = (word in document_words)
  return features


  #returns a dictionary of whether the word in a document has words in the word_feature list 

In [578]:
print(document_features(movie_reviews.words('pos/cv957_8737.txt'))) 
#so the text file contains the word plot .... doesn't contain the word couples 

{'plot': True, ':': True, 'two': True, 'teen': False, 'couples': False, 'go': False, 'to': True, 'a': True, 'church': False, 'party': False, ',': True, 'drink': False, 'and': True, 'then': True, 'drive': False, '.': True, 'they': True, 'get': True, 'into': True, 'an': True, 'accident': False, 'one': True, 'of': True, 'the': True, 'guys': False, 'dies': False, 'but': True, 'his': True, 'girlfriend': True, 'continues': False, 'see': False, 'him': True, 'in': True, 'her': False, 'life': False, 'has': True, 'nightmares': False, 'what': True, "'": True, 's': True, 'deal': False, '?': False, 'watch': True, 'movie': True, '"': True, 'sorta': False, 'find': False, 'out': True, 'critique': False, 'mind': False, '-': True, 'fuck': False, 'for': True, 'generation': False, 'that': True, 'touches': False, 'on': True, 'very': True, 'cool': False, 'idea': True, 'presents': False, 'it': True, 'bad': False, 'package': False, 'which': True, 'is': True, 'makes': False, 'this': True, 'review': False, 'eve

In [0]:
featuresets = [(document_features(d), c) for (d,c) in documents]
#featuresets is a list of whether the document includes the words inside word_features 
#featuresets[0] #the first document has words of plot, to, a, .... and is a negative movie review 

In [580]:
#divide the featuresets to train/test (assign 100 test sets, others as train set)
train, test = featuresets[100:], featuresets[:100]
len(train)#1900 
len(test)#100

100

In [581]:
classifier = nltk.NaiveBayesClassifier.train(train) #train with NBClassifier 
first_classifier_accuracy =nltk.classify.accuracy(classifier,test)
print(nltk.classify.accuracy(classifier,test)) #test with the accuracy 

0.76


In [582]:
classifier.show_most_informative_features(10)

Most Informative Features
                  justin = True              neg : pos    =      8.3 : 1.0
           unimaginative = True              neg : pos    =      8.3 : 1.0
              schumacher = True              neg : pos    =      7.4 : 1.0
                  suvari = True              neg : pos    =      7.0 : 1.0
                    mena = True              neg : pos    =      7.0 : 1.0
                  shoddy = True              neg : pos    =      7.0 : 1.0
               atrocious = True              neg : pos    =      7.0 : 1.0
                  turkey = True              neg : pos    =      6.8 : 1.0
                 singers = True              pos : neg    =      6.3 : 1.0
                 jumbled = True              neg : pos    =      6.3 : 1.0


## **Now lets improve the classifier!!**



In [583]:
from nltk.corpus import stopwords #high-frequency words that we want to filter out of a document 
stopwords.words('english')[:10]

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

In [584]:

new_all_words = {}
for  key in all_words.keys():
  if key not in stopwords.words('english'):
    new_all_words[key] = all_words[key]

len(new_all_words) #about 100 words are deleted 


39617

In [585]:
sorted(new_all_words.items(), key = lambda item: item[1], reverse=True)[:30] #delete .,'"_)()
#?! may have effect on the pos/neg of movie review so delete only 

[(',', 77717),
 ('.', 65876),
 ("'", 30585),
 ('"', 17612),
 ('-', 15595),
 (')', 11781),
 ('(', 11664),
 ('film', 9517),
 ('one', 5852),
 ('movie', 5771),
 ('?', 3771),
 ('like', 3690),
 (':', 3042),
 ('even', 2565),
 ('good', 2411),
 ('time', 2411),
 ('story', 2169),
 ('would', 2109),
 ('much', 2049),
 ('character', 2020),
 ('also', 1967),
 ('get', 1949),
 ('two', 1911),
 ('well', 1906),
 ('characters', 1859),
 (';', 1850),
 ('first', 1836),
 ('--', 1815),
 ('see', 1749),
 ('!', 1713)]

In [586]:
#update from new_all_words dictionary 
llist = [',','.','(',')',"'",'"','_','-',':']

update_all_words = {}
for  key in new_all_words.keys():
  if key not in llist:
    update_all_words[key] = new_all_words[key]

len(update_all_words) 


39608

In [587]:
word_sort2 = sorted(update_all_words.items(), key = lambda item: item[1], reverse=True) 
word_sort2[:30]


[('film', 9517),
 ('one', 5852),
 ('movie', 5771),
 ('?', 3771),
 ('like', 3690),
 ('even', 2565),
 ('good', 2411),
 ('time', 2411),
 ('story', 2169),
 ('would', 2109),
 ('much', 2049),
 ('character', 2020),
 ('also', 1967),
 ('get', 1949),
 ('two', 1911),
 ('well', 1906),
 ('characters', 1859),
 (';', 1850),
 ('first', 1836),
 ('--', 1815),
 ('see', 1749),
 ('!', 1713),
 ('way', 1693),
 ('make', 1642),
 ('life', 1586),
 ('really', 1558),
 ('films', 1536),
 ('plot', 1513),
 ('little', 1501),
 ('people', 1455)]

In [588]:
second_word_features= []
for word in word_sort2:
  second_word_features.append(word[0]) #make a list of words from the sorted list 

second_word_features[:10]


['film', 'one', 'movie', '?', 'like', 'even', 'good', 'time', 'story', 'would']

In [589]:

def second_document_features(document):
  document_words = set(document)
  features = {}
  for word in second_word_features[:3000]: 
    #print(word)a
    features[word] = (word in document_words)

  return features

print(second_document_features(movie_reviews.words('pos/cv957_8737.txt'))) 
#randomly[:10]

{'film': False, 'one': True, 'movie': True, '?': False, 'like': True, 'even': False, 'good': False, 'time': False, 'story': False, 'would': False, 'much': False, 'character': False, 'also': True, 'get': True, 'two': True, 'well': True, 'characters': False, ';': False, 'first': False, '--': True, 'see': False, '!': True, 'way': True, 'make': True, 'life': False, 'really': False, 'films': False, 'plot': True, 'little': True, 'people': False, 'could': False, 'scene': True, 'man': False, 'bad': False, 'never': True, 'best': True, 'new': False, 'scenes': False, 'many': True, 'director': False, 'know': False, 'movies': True, 'action': True, 'great': True, 'another': False, 'love': False, 'go': False, 'made': False, 'us': True, 'big': False, 'end': False, 'something': False, 'back': False, '*': True, 'still': False, 'world': True, 'seems': False, 'work': False, 'makes': False, 'however': True, '/': False, 'every': False, 'though': False, 'better': False, 'real': False, 'audience': False, 'eno

In [0]:
second_featuresets = [(second_document_features(d), c) for (d,c) in documents]
second_train, second_test = second_featuresets[100:], second_featuresets[:100]

In [591]:
second_classifier = nltk.NaiveBayesClassifier.train(second_train) #train with NBClassifier 
second_classifier_accuracy =nltk.classify.accuracy(second_classifier,second_test)
print(second_classifier_accuracy) #test with the accuracy 

#still doesn't have good accuracy 

0.82


In [592]:
second_classifier.show_most_informative_features(50)

Most Informative Features
             outstanding = True              pos : neg    =     10.8 : 1.0
                   jolie = True              neg : pos    =      9.0 : 1.0
                   mulan = True              pos : neg    =      9.0 : 1.0
                  seagal = True              neg : pos    =      7.8 : 1.0
              schumacher = True              neg : pos    =      7.4 : 1.0
                  finest = True              pos : neg    =      7.2 : 1.0
             wonderfully = True              pos : neg    =      6.6 : 1.0
            breathtaking = True              pos : neg    =      6.5 : 1.0
                   anger = True              pos : neg    =      6.3 : 1.0
                 idiotic = True              neg : pos    =      6.1 : 1.0
                  random = True              neg : pos    =      5.7 : 1.0
                   flynt = True              pos : neg    =      5.7 : 1.0
                lebowski = True              pos : neg    =      5.7 : 1.0

In [593]:
#lets delete all the names : names don't have positive/negative effects on the review 

from nltk.corpus import names 
male = names.words('male.txt') #list of male names 
female = names.words('female.txt') #list of female names 

malelist=[]
for name in male:
  malelist.append(name.lower())

femalelist=[]
for name in female:
  femalelist.append(name.lower())

namelist = malelist+femalelist

namelist[:10]


['aamir',
 'aaron',
 'abbey',
 'abbie',
 'abbot',
 'abbott',
 'abby',
 'abdel',
 'abdul',
 'abdulkarim']

In [594]:
update_words = {}
for  key in update_all_words.keys():
    if key not in namelist:
        update_words[key] = update_all_words[key]

len(update_words) #about 2000 words deleted 


37219

In [595]:
word_sort = sorted(update_words.items(), key = lambda item: item[1], reverse=True) #use the sorted list to select 2000 words 
new_word_features= []
for word in word_sort:
  new_word_features.append(word[0])

new_word_features[:10]


['film', 'one', 'movie', '?', 'like', 'even', 'good', 'time', 'story', 'would']

In [596]:

def third_document_features(document):
  document_words = set(document)
  features = {}
  for word in new_word_features[:3000]:
    features[word] = (word in document_words)
  return features

print(third_document_features(movie_reviews.words('pos/cv957_8737.txt')))

{'film': False, 'one': True, 'movie': True, '?': False, 'like': True, 'even': False, 'good': False, 'time': False, 'story': False, 'would': False, 'much': False, 'character': False, 'also': True, 'get': True, 'two': True, 'well': True, 'characters': False, ';': False, 'first': False, '--': True, '!': True, 'make': True, 'life': False, 'really': False, 'films': False, 'plot': True, 'little': True, 'people': False, 'could': False, 'scene': True, 'man': False, 'bad': False, 'never': True, 'best': True, 'new': False, 'scenes': False, 'many': True, 'director': False, 'know': False, 'movies': True, 'action': True, 'great': True, 'another': False, 'go': False, 'made': False, 'us': True, 'big': False, 'end': False, 'something': False, 'back': False, '*': True, 'still': False, 'world': True, 'seems': False, 'work': False, 'makes': False, 'however': True, '/': False, 'every': False, 'though': False, 'better': False, 'real': False, 'audience': False, 'enough': False, 'seen': False, 'take': False,

In [597]:
third_featuresets = [(third_document_features(d), c) for (d,c) in documents]
third_train, third_test = third_featuresets[100:], third_featuresets[:100]
third_classifier = nltk.NaiveBayesClassifier.train(third_train) #train with NBClassifier 
third_classifier_accuracy =nltk.classify.accuracy(third_classifier,third_test)
print(nltk.classify.accuracy(third_classifier,third_test)) 

0.78


In [598]:
classifier.show_most_informative_features(10)

Most Informative Features
                  justin = True              neg : pos    =      8.3 : 1.0
           unimaginative = True              neg : pos    =      8.3 : 1.0
              schumacher = True              neg : pos    =      7.4 : 1.0
                  suvari = True              neg : pos    =      7.0 : 1.0
                    mena = True              neg : pos    =      7.0 : 1.0
                  shoddy = True              neg : pos    =      7.0 : 1.0
               atrocious = True              neg : pos    =      7.0 : 1.0
                  turkey = True              neg : pos    =      6.8 : 1.0
                 singers = True              pos : neg    =      6.3 : 1.0
                 jumbled = True              neg : pos    =      6.3 : 1.0


In [599]:
print(first_classifier_accuracy)
print(second_classifier_accuracy)
print(third_classifier_accuracy)

0.76
0.82
0.78


The second classifier has the best accuracy, not deleting the names is better: return to second classifier 

In [600]:
errors = [] 

for (word, expected) in second_test:
  guess = classifier.classify(second_document_features(word))

  if guess != tag:
    errors.append((expected,guess,word))

errors[0] #1: expected 2: guessed as.. 3: words included in the error


('neg',
 'pos',
 {'film': True,
  'one': True,
  'movie': False,
  '?': True,
  'like': False,
  'even': False,
  'good': False,
  'time': True,
  'story': True,
  'would': True,
  'much': False,
  'character': False,
  'also': False,
  'get': False,
  'two': True,
  'well': False,
  'characters': True,
  ';': False,
  'first': True,
  '--': False,
  'see': True,
  '!': True,
  'way': False,
  'make': False,
  'life': False,
  'really': False,
  'films': False,
  'plot': False,
  'little': False,
  'people': True,
  'could': False,
  'scene': False,
  'man': False,
  'bad': False,
  'never': False,
  'best': False,
  'new': True,
  'scenes': False,
  'many': False,
  'director': False,
  'know': False,
  'movies': True,
  'action': False,
  'great': False,
  'another': False,
  'love': False,
  'go': False,
  'made': False,
  'us': False,
  'big': False,
  'end': False,
  'something': False,
  'back': False,
  '*': False,
  'still': False,
  'world': False,
  'seems': False,
  'work': 

Back to Second version:
Lets not use True or false , just use total numbers of occurance 


In [601]:
update_all_words['plot']

1513

In [616]:
sorted_words = sorted(update_all_words.items(), key = lambda item: item[1], reverse=True)
top3000 = sorted_words[:3000]

top3000_dict = {}
for item in top3000:
  top3000_dict[item[0]] = item[1]


len(top3000_dict)

3000

In [623]:


def second_document_features(document):
  document_words = document
  fdist = nltk.FreqDist(document)
  #print(fdist.most_common(10))
  features ={}
  for key in top3000_dict.keys():
    if key in fdist.keys():
      features[key] = fdist[key]
    else:
      features[key] = 0
    

  return features

print(second_document_features(movie_reviews.words('pos/cv957_8737.txt'))) 


{'film': 0, 'one': 4, 'movie': 3, '?': 0, 'like': 1, 'even': 0, 'good': 0, 'time': 0, 'story': 0, 'would': 0, 'much': 0, 'character': 0, 'also': 4, 'get': 1, 'two': 2, 'well': 1, 'characters': 0, ';': 0, 'first': 0, '--': 5, 'see': 0, '!': 1, 'way': 4, 'make': 1, 'life': 0, 'really': 0, 'films': 0, 'plot': 1, 'little': 1, 'people': 0, 'could': 0, 'scene': 2, 'man': 0, 'bad': 0, 'never': 1, 'best': 1, 'new': 0, 'scenes': 0, 'many': 1, 'director': 0, 'know': 0, 'movies': 4, 'action': 4, 'great': 1, 'another': 0, 'love': 0, 'go': 0, 'made': 0, 'us': 1, 'big': 0, 'end': 0, 'something': 0, 'back': 0, '*': 2, 'still': 0, 'world': 1, 'seems': 0, 'work': 0, 'makes': 0, 'however': 1, '/': 0, 'every': 0, 'though': 0, 'better': 0, 'real': 0, 'audience': 0, 'enough': 0, 'seen': 0, 'take': 0, 'around': 0, 'going': 0, 'year': 0, 'performance': 0, 'role': 0, 'old': 0, 'gets': 1, 'may': 0, 'things': 2, 'think': 0, 'years': 0, 'last': 0, 'comedy': 1, 'funny': 2, 'actually': 1, 'long': 0, 'look': 1, 'al

In [0]:
second_featuresets = [(second_document_features(d), c) for (d,c) in documents]

In [625]:
second_featuresets[0]

({'film': 1,
  'one': 3,
  'movie': 0,
  '?': 1,
  'like': 0,
  'even': 0,
  'good': 0,
  'time': 1,
  'story': 1,
  'would': 2,
  'much': 0,
  'character': 0,
  'also': 0,
  'get': 0,
  'two': 1,
  'well': 0,
  'characters': 1,
  ';': 0,
  'first': 1,
  '--': 0,
  'see': 1,
  '!': 1,
  'way': 0,
  'make': 0,
  'life': 0,
  'really': 0,
  'films': 0,
  'plot': 0,
  'little': 0,
  'people': 1,
  'could': 0,
  'scene': 0,
  'man': 0,
  'bad': 0,
  'never': 0,
  'best': 0,
  'new': 1,
  'scenes': 0,
  'many': 0,
  'director': 0,
  'know': 0,
  'movies': 1,
  'action': 0,
  'great': 0,
  'another': 0,
  'love': 0,
  'go': 0,
  'made': 0,
  'us': 0,
  'big': 0,
  'end': 0,
  'something': 0,
  'back': 0,
  '*': 0,
  'still': 0,
  'world': 0,
  'seems': 0,
  'work': 0,
  'makes': 0,
  'however': 0,
  '/': 0,
  'every': 0,
  'though': 0,
  'better': 0,
  'real': 0,
  'audience': 0,
  'enough': 0,
  'seen': 1,
  'take': 1,
  'around': 0,
  'going': 0,
  'year': 0,
  'performance': 0,
  'role': 

In [626]:

second_train, second_test = second_featuresets[100:], second_featuresets[:100]

second_classifier = nltk.NaiveBayesClassifier.train(second_train) #train with NBClassifier 
new_second_classifier_accuracy =nltk.classify.accuracy(second_classifier,second_test)

print(new_second_classifier_accuracy) #test with the accuracy 


0.83


Lets upgrade the accuracy by using laplace smoothing (similar; not added total unique) 

In [627]:


def new_second_document_features(document):
  document_words = document
  fdist = nltk.FreqDist(document)
  #print(fdist.most_common(10))
  features ={}
  for key in top3000_dict.keys():
    if key in fdist.keys():
      features[key] = fdist[key]+ 1 
    else:
      features[key] = 1
    #add one to make non-zeros 

  return features

second_featuresets = [(new_second_document_features(d), c) for (d,c) in documents]

second_train, second_test = second_featuresets[100:], second_featuresets[:100]

second_classifier = nltk.NaiveBayesClassifier.train(second_train) #train with NBClassifier 
laplace_second_classifier_accuracy =nltk.classify.accuracy(second_classifier,second_test)

print(laplace_second_classifier_accuracy) #test with the accuracy 


0.83


In [628]:
second_classifier.show_most_informative_features(20)

Most Informative Features
               wonderful = 3                 pos : neg    =     14.3 : 1.0
              apparently = 3                 neg : pos    =     13.7 : 1.0
                  stupid = 3                 neg : pos    =     12.2 : 1.0
                  boring = 3                 neg : pos    =     11.8 : 1.0
                     bad = 6                 neg : pos    =     10.3 : 1.0
             outstanding = 2                 pos : neg    =     10.1 : 1.0
                     bad = 7                 neg : pos    =      8.3 : 1.0
                 country = 3                 pos : neg    =      8.3 : 1.0
            breathtaking = 2                 pos : neg    =      7.7 : 1.0
                provides = 3                 pos : neg    =      7.7 : 1.0
               memorable = 3                 pos : neg    =      7.7 : 1.0
                  parody = 3                 pos : neg    =      7.0 : 1.0
                    crap = 3                 neg : pos    =      7.0 : 1.0

In [629]:
print(first_classifier_accuracy)
print(second_classifier_accuracy)
print(third_classifier_accuracy)
print(new_second_classifier_accuracy)
print(laplace_second_classifier_accuracy)

0.76
0.82
0.78
0.83
0.83


Looks like the laplace_second_Classifier has the best accuracy 
Improved the accuracy to .76->.83 

